File: impl/kotlin/src/test/kotlin/AuditTests.kt
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
package sh.bubblefish.naalp

import java.io.File
import java.security.MessageDigest
import kotlin.system.exitProcess

/**
 * C7 audit known-answer test for the Kotlin SDK (design.md §8; R-8.1..8.6, R-12.2, R-12.3), graded
 * against the shared independent corpus vectors/audit/cases.json (NOT produced by this code).
 *
 * CORPUS-GRADED (pure): the hash-chained signed-receipt body + head for each receipt, the running
 * prev-link and final head, the draft-01 fork-proof preimage (both signatures elided to empty), the
 * equivocation receipt bodies, and the offline causal graph (valid topo order, cycle rejection,
 * future-cause rejection). ISOLATION (real FIPS-204 ML-DSA-65 via BouncyCastle — Kotlin is a full-
 * signature port; the corpus carries no signature vector for this channel, so these are demonstrated
 * with a fixed local seed, stated honestly, NOT corpus-graded): a fresh Authority reproduces the
 * byte-exact chain and its real-ML-DSA-signed receipts verify offline; a broken prev-link is
 * ChainBroken; a tampered signature is ReceiptUnsigned; the auditor mints a verifying fork proof on a
 * genuine one-seq conflict and rejects a benign duplicate / an unsigned receipt; the fork proof
 * fails closed (ForkProofInvalid / ReceiptUnsigned).
 *
 * KAT convention: a standalone main() exiting non-zero on any failure; the filename carries the
 * "Tests" token the ten-language-parity gate indexes. Written test-first: [Audit] is absent until
 * Audit.kt lands, so this fails RED with a kotlinc "unresolved reference: Audit". A mutation that
 * forces the encoded receipt seq field to a constant flips "chain receipt body seq=1 == oracle".
 */

private var auFails = 0

private fun auCheck(name: String, got: String, want: String) {
    if (got == want) {
        println("  ok   $name")
    } else {
        auFails++
        println("  FAIL $name\n       got  $got\n       want $want")
    }
}

// ---- balanced-brace / regex JSON access (no JSON library on the Kotlin port) ----

private fun auFindVector(): File {
    var d: File? = File(".").absoluteFile
    repeat(6) {
        val cur = d ?: throw AssertionError("vectors/audit/cases.json not found")
        val p = File(File(cur, "vectors"), "audit/cases.json")
        if (p.isFile) return p
        d = cur.parentFile
    }
    throw AssertionError("vectors/audit/cases.json not found from ${File(".").absolutePath}")
}

/**
 * The balanced { ... } or [ ... ] value text (delimiters included) that follows "key":, scanning
 * with brace/bracket depth while skipping string contents. Robust to nesting (arrays inside objects).
 */
private fun auSection(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*").find(scope)
        ?: throw AssertionError("key not found: $key")
    var i = m.range.last + 1
    while (i < scope.length && scope[i].isWhitespace()) i++
    val open = scope[i]
    val close = when (open) { '{' -> '}'; '[' -> ']'; else -> throw AssertionError("not object/array at $key") }
    var depth = 0
    var inStr = false
    var esc = false
    val start = i
    while (i < scope.length) {
        val c = scope[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                open -> depth++
                close -> { depth--; if (depth == 0) return scope.substring(start, i + 1) }
            }
        }
        i++
    }
    throw AssertionError("unbalanced value at key: $key")
}

/** Top-level { ... } object element substrings of an already-extracted [ ... ] array section. */
private fun auElements(arraySection: String): List<String> {
    val out = ArrayList<String>()
    var i = 1
    var inStr = false
    var esc = false
    var depth = 0
    var start = -1
    while (i < arraySection.length - 1) {
        val c = arraySection[i]
        if (inStr) {
            when {
                esc -> esc = false
                c == '\\' -> esc = true
                c == '"' -> inStr = false
            }
        } else {
            when (c) {
                '"' -> inStr = true
                '{' -> { if (depth == 0) start = i; depth++ }
                '}' -> { depth--; if (depth == 0 && start >= 0) { out.add(arraySection.substring(start, i + 1)); start = -1 } }
            }
        }
        i++
    }
    return out
}

/** The string value of a "key": "value" pair inside [scope] (value may be empty). */
private fun auStr(scope: String, key: String): String {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*\"([^\"]*)\"").find(scope)
        ?: throw AssertionError("string key not found: $key")
    return m.groupValues[1]
}

/** The integer value of a "key": <number> pair inside [scope]. */
private fun auInt(scope: String, key: String): Long {
    val m = Regex("\"" + Regex.escape(key) + "\"\\s*:\\s*(\\d+)").find(scope)
        ?: throw AssertionError("int key not found: $key")
    return m.groupValues[1].toLong()
}

/** The list of quoted hex strings of the array-valued field named [key] inside [scope]. */
private fun auHexList(scope: String, key: String): List<String> {
    val arr = auSection(scope, key)
    return Regex("\"([0-9a-fA-F]*)\"").findAll(arr).map { it.groupValues[1] }.toList()
}

private fun auHex(s: String): ByteArray = Hex.decode(s)
private fun auSha384Hex(b: ByteArray): String = Hex.encode(MessageDigest.getInstance("SHA-384").digest(b))

private const val AU_ALG = Cose.ALG_MLDSA65
private val AU_SEED = ByteArray(32)
private val AU_PK = Cose.mldsaKeygen("ML-DSA-65", ByteArray(32))

/** The named kind thrown by [block], or "no-error". */
private inline fun auKind(block: () -> Unit): String = try {
    block(); "no-error"
} catch (e: NaalpException) {
    e.kind
}

private fun auReceipt(block: String): Audit.Receipt =
    Audit.Receipt(auHex(auStr(block, "prev_hex")), auHex(auStr(block, "obj_hex")), auInt(block, "seq"), auInt(block, "at"))

private fun auCausalNodes(scope: String): List<Audit.CausalNode> =
    auElements(auSection(scope, "nodes")).map { nb ->
        Audit.CausalNode(auHex(auStr(nb, "id_hex")), auHexList(nb, "causes_hex").map { auHex(it) }, auInt(nb, "position"))
    }

private fun auditRun() {
    val json = auFindVector().readText(Charsets.UTF_8)

    // 1. signed hash-chained receipt: each receipt body/head == the oracle, and prev links the chain.
    val chain = auSection(json, "chain")
    var head = auHex(auStr(chain, "genesis_prev_hex"))
    auCheck("genesis prev is HEAD_SIZE zeros", head.size.toString() + "/" + Hex.encode(head), "48/" + "00".repeat(48))
    auCheck("HEAD_SIZE", Audit.HEAD_SIZE.toString(), "48")
    for (rb in auElements(auSection(chain, "receipts"))) {
        val r = auReceipt(rb)
        auCheck("chain receipt body seq=${r.seq} == oracle", Hex.encode(r.bytes()), auStr(rb, "body_hex"))
        auCheck("chain receipt head seq=${r.seq} == oracle", Hex.encode(r.head()), auStr(rb, "head_after_hex"))
        auCheck("chain prev links seq=${r.seq}", Hex.encode(r.prev), Hex.encode(head))
        head = r.head()
    }
    auCheck("chain final head == oracle", Hex.encode(head), auStr(chain, "final_head_hex"))

    // 2. a fresh authority reproduces the byte-exact chain; its real-ML-DSA-signed receipts verify
    //    offline (ISOLATION — corpus has no signature vector).
    run {
        val auth = Audit.Authority(AU_ALG, AU_SEED)
        val receipts = ArrayList<Audit.Receipt>()
        val sigs = ArrayList<ByteArray>()
        for (rb in auElements(auSection(chain, "receipts"))) {
            val (r, sig) = auth.append(auHex(auStr(rb, "obj_hex")), auInt(rb, "at"))
            auCheck("authority append body seq=${r.seq} == oracle", Hex.encode(r.bytes()), auStr(rb, "body_hex"))
            receipts.add(r); sigs.add(sig)
        }
        auCheck("authority chain verifies offline", auKind { Audit.verifyChain(receipts, sigs, AU_ALG, AU_PK) }, "no-error")
    }

    // 3. a broken prev-link is ChainBroken (each body signed with a real key so the break, not a bad
    //    signature, is what fires).
    run {
        val cb = auSection(json, "chain_broken")
        val receipts = ArrayList<Audit.Receipt>()
        val sigs = ArrayList<ByteArray>()
        for (rb in auElements(auSection(cb, "receipts"))) {
            val r = auReceipt(rb)
            auCheck("chain_broken body seq=${r.seq} == oracle", Hex.encode(r.bytes()), auStr(rb, "body_hex"))
            receipts.add(r); sigs.add(Cose.mldsaSign(AU_ALG, AU_SEED, r.bytes()))
        }
        auCheck("broken prev-link rejected", auKind { Audit.verifyChain(receipts, sigs, AU_ALG, AU_PK) }, auStr(cb, "expect"))
    }

    // 4. a tampered signature is ReceiptUnsigned (ISOLATION).
    run {
        val rb = auElements(auSection(chain, "receipts"))[0]
        val r = Audit.Receipt(auHex(auStr(chain, "genesis_prev_hex")), auHex(auStr(rb, "obj_hex")), 0, auInt(rb, "at"))
        val bad = Cose.mldsaSign(AU_ALG, AU_SEED, r.bytes()).copyOf()
        bad[bad.size - 1] = (bad[bad.size - 1].toInt() xor 1).toByte()
        auCheck("tampered signature rejected", auKind { Audit.verifyChain(listOf(r), listOf(bad), AU_ALG, AU_PK) }, "ReceiptUnsigned")
    }

    // 5. consistent-with-anchor: an object cannot be created after the authority ordered it.
    auCheck("anchor created==at ok", Audit.consistentWithAnchor(100, 100).toString(), "true")
    auCheck("anchor created<at ok", Audit.consistentWithAnchor(99, 100).toString(), "true")
    auCheck("anchor created>at rejected", Audit.consistentWithAnchor(101, 100).toString(), "false")

    // 6. equivocation receipts (two receipts at one seq naming different objects) match the oracle.
    val fp = auSection(json, "fork_proof")
    val ra = Audit.Receipt(auHex(auStr(fp, "prev_hex")), auHex(auStr(fp, "obj_a_hex")), auInt(fp, "seq"), auInt(fp, "at"))
    val rb = Audit.Receipt(auHex(auStr(fp, "prev_hex")), auHex(auStr(fp, "obj_b_hex")), auInt(fp, "seq"), auInt(fp, "at"))
    auCheck("fork receipt A body == oracle", Hex.encode(ra.bytes()), auStr(fp, "body_a_hex"))
    auCheck("fork receipt B body == oracle", Hex.encode(rb.bytes()), auStr(fp, "body_b_hex"))
    val eq = auSection(json, "equivocation")
    auCheck("equivocation receipt_a body == oracle", Hex.encode(ra.bytes()), auStr(auSection(eq, "receipt_a"), "body_hex"))
    auCheck("equivocation receipt_b body == oracle", Hex.encode(rb.bytes()), auStr(auSection(eq, "receipt_b"), "body_hex"))

    // 7. the auditor detects a genuine one-seq fork and mints a verifying proof; a benign duplicate
    //    and an unsigned receipt do not (ISOLATION).
    run {
        val signer = auHex(auStr(fp, "signer_hex"))
        val ext = auInt(fp, "ext_counter")
        val sigA = Cose.mldsaSign(AU_ALG, AU_SEED, ra.bytes())
        val sigB = Cose.mldsaSign(AU_ALG, AU_SEED, rb.bytes())
        val auditor = Audit.Auditor(AU_ALG, AU_PK, signer, ext)
        auCheck("first observe benign", (auditor.observe(ra, sigA) == null).toString(), "true")
        val proof = auditor.observe(rb, sigB)
        auCheck("second observe mints proof", (proof != null).toString(), "true")
        auCheck("minted proof verifies", auKind { proof!!.verify(AU_ALG, AU_PK) }, "no-error")
        auCheck("corpus equivocation expect", auStr(eq, "expect"), "Equivocation")

        val dupAuditor = Audit.Auditor(AU_ALG, AU_PK, signer, ext)
        auCheck("duplicate observe 1 benign", (dupAuditor.observe(ra, sigA) == null).toString(), "true")
        auCheck("exact duplicate is not a fork", (dupAuditor.observe(ra, sigA) == null).toString(), "true")

        val foreignSig = Cose.mldsaSign(AU_ALG, ByteArray(32) { 0x77 }, ra.bytes()) // valid-length, verifies false
        auCheck("unsigned receipt rejected", auKind { Audit.Auditor(AU_ALG, AU_PK, signer, ext).observe(ra, foreignSig) }, "ReceiptUnsigned")
    }

    // 8. the fork-proof framing witness (both signatures elided to empty) == the oracle preimage.
    run {
        val signer = auHex(auStr(fp, "signer_hex"))
        val ext = auInt(fp, "ext_counter")
        val proof = Audit.newForkProof(signer, ra, ByteArray(0), rb, ByteArray(0), ext)
        auCheck("fork-proof preimage == oracle", Hex.encode(proof.preimage()), auStr(fp, "preimage_hex"))
    }

    // 9. the fork proof accepts a valid proof and fails closed on every degenerate shape (ISOLATION).
    run {
        val signer = auHex(auStr(fp, "signer_hex"))
        val ext = auInt(fp, "ext_counter")
        val sigA = Cose.mldsaSign(AU_ALG, AU_SEED, ra.bytes())
        val sigB = Cose.mldsaSign(AU_ALG, AU_SEED, rb.bytes())
        auCheck("valid fork proof accepts", auKind { Audit.newForkProof(signer, ra, sigA, rb, sigB, ext).verify(AU_ALG, AU_PK) }, "no-error")
        auCheck("same-object proof rejected", auKind { Audit.newForkProof(signer, ra, sigA, ra, sigA, ext).verify(AU_ALG, AU_PK) }, "ForkProofInvalid")
        auCheck("unnamed-signer proof rejected", auKind { Audit.newForkProof(ByteArray(0), ra, sigA, rb, sigB, ext).verify(AU_ALG, AU_PK) }, "ForkProofInvalid")
        val rbSeq = Audit.Receipt(rb.prev, rb.obj, rb.seq + 1, rb.at)
        val sigB2 = Cose.mldsaSign(AU_ALG, AU_SEED, rbSeq.bytes())
        auCheck("seq-mismatch proof rejected", auKind { Audit.newForkProof(signer, ra, sigA, rbSeq, sigB2, ext).verify(AU_ALG, AU_PK) }, "ForkProofInvalid")
        val badSig = sigB.copyOf(); badSig[badSig.size - 1] = (badSig[badSig.size - 1].toInt() xor 1).toByte()
        auCheck("tampered-signature proof rejected", auKind { Audit.newForkProof(signer, ra, sigA, rb, badSig, ext).verify(AU_ALG, AU_PK) }, "ReceiptUnsigned")
    }

    // 10. offline causal graph: valid topo order == oracle; cycle and future-cause are CausalViolation.
    run {
        val cv = auSection(json, "causal_valid")
        val nodes = auCausalNodes(cv)
        auCheck("causal valid verifies", auKind { Audit.verifyCausal(nodes) }, "no-error")
        auCheck("causal topo order == oracle", Audit.topoOrder(nodes).joinToString(",") { Hex.encode(it) }, auHexList(cv, "topo_order_hex").joinToString(","))
        val cc = auSection(json, "causal_cycle")
        auCheck("causal cycle rejected", auKind { Audit.verifyCausal(auCausalNodes(cc)) }, auStr(cc, "expect"))
        val cf = auSection(json, "causal_future")
        auCheck("causal future-cause rejected", auKind { Audit.verifyCausal(auCausalNodes(cf)) }, auStr(cf, "expect"))
    }
}

fun main() {
    println("audit conformance (Kotlin) — graded vs vectors/audit/cases.json")
    auditRun()
    println(if (auFails == 0) "AuditTests: PASS" else "AuditTests: FAIL ($auFails)")
    exitProcess(if (auFails == 0) 0 else 1)
}

File: impl/csharp/test/AuditTests.cs
// Copyright (c) 2026 BubbleFish Technologies, Inc. Apache-2.0.
using System;
using System.Collections.Generic;
using System.IO;
using System.Text.Json;

using Naalp;
using Xunit;

namespace Sh.Bubblefish.Naalp.Tests
{
    /// <summary>
    /// C7 audit conformance for the C# SDK, graded against the shared independent corpus
    /// <c>vectors/audit/cases.json</c> (NOT produced by this code): the hash-chained signed receipt
    /// body + head, offline chain verification (ChainBroken / ReceiptUnsigned), equivocation
    /// detection, the draft-01 fork-proof preimage (signatures elided), and the offline causal graph
    /// (valid topo order, cycle rejection, future-cause rejection).
    ///
    /// <para>The receipt/fork-proof SIGNATURES are real deterministic ML-DSA-65 (BouncyCastle, rnd=0),
    /// but the corpus carries no signature vector for this channel (signatures are graded by the shared
    /// cose byte-parity — the worked-example / gateway pins), so the sign/verify round-trips here are
    /// demonstrated in isolation with a fixed local seed — stated honestly, not corpus-graded. Every
    /// byte-exact assertion (body/head/preimage/topo) IS corpus-graded. Ported from impl/go/audit;
    /// mirrors impl/python/tests/test_audit.py.</para>
    /// </summary>
    public sealed class AuditTests
    {
        private const int Alg = Cose.ALG_MLDSA65;
        private static readonly byte[] Seed = new byte[32];
        private static readonly byte[] Pk = Cose.MldsaKeygen("ML-DSA-65", Seed);

        private static JsonElement Vector()
        {
            var d = new DirectoryInfo(AppContext.BaseDirectory);
            for (int i = 0; i < 12 && d != null; i++)
            {
                string p = Path.Combine(d.FullName, "vectors", "audit", "cases.json");
                if (File.Exists(p))
                {
                    return JsonDocument.Parse(File.ReadAllText(p)).RootElement.Clone();
                }
                d = d.Parent;
            }
            throw new Xunit.Sdk.XunitException("vectors/audit/cases.json not found from " + AppContext.BaseDirectory);
        }

        private static string Hex(byte[] b) => Convert.ToHexString(b).ToLowerInvariant();
        private static byte[] Hb(string s) => Convert.FromHexString(s);

        // ---- signed hash-chained receipt (design §8.1) ----------------------------------------

        [Fact]
        public void ChainReceiptsMatchOracle()
        {
            JsonElement v = Vector();
            JsonElement chain = v.GetProperty("chain");
            byte[] head = Hb(chain.GetProperty("genesis_prev_hex").GetString()!);
            Assert.Equal(Audit.HeadSize, head.Length);
            foreach (JsonElement rv in chain.GetProperty("receipts").EnumerateArray())
            {
                var r = new Audit.Receipt(
                    Hb(rv.GetProperty("prev_hex").GetString()!),
                    Hb(rv.GetProperty("obj_hex").GetString()!),
                    rv.GetProperty("seq").GetInt64(),
                    rv.GetProperty("at").GetInt64());
                Assert.Equal(rv.GetProperty("body_hex").GetString(), Hex(r.Bytes()));
                Assert.Equal(rv.GetProperty("head_after_hex").GetString(), Hex(r.Head()));
                Assert.Equal(Hex(head), Hex(r.Prev));
                head = r.Head();
            }
            Assert.Equal(chain.GetProperty("final_head_hex").GetString(), Hex(head));
        }

        [Fact]
        public void AuthorityReproducesChainAndVerifies()
        {
            JsonElement v = Vector();
            var authority = new Audit.Authority(Alg, Seed);
            var receipts = new List<Audit.Receipt>();
            var sigs = new List<byte[]>();
            foreach (JsonElement rv in v.GetProperty("chain").GetProperty("receipts").EnumerateArray())
            {
                (Audit.Receipt r, byte[] sig) = authority.Append(
                    Hb(rv.GetProperty("obj_hex").GetString()!), rv.GetProperty("at").GetInt64());
                Assert.Equal(rv.GetProperty("body_hex").GetString(), Hex(r.Bytes()));
                receipts.Add(r);
                sigs.Add(sig);
            }
            Audit.VerifyChain(receipts, sigs, Alg, Pk); // real-ML-DSA chain verifies offline
        }

        [Fact]
        public void VerifyChainDetectsBrokenPrevLink()
        {
            JsonElement v = Vector();
            JsonElement cb = v.GetProperty("chain_broken");
            var receipts = new List<Audit.Receipt>();
            var sigs = new List<byte[]>();
            foreach (JsonElement rv in cb.GetProperty("receipts").EnumerateArray())
            {
                var r = new Audit.Receipt(
                    Hb(rv.GetProperty("prev_hex").GetString()!),
                    Hb(rv.GetProperty("obj_hex").GetString()!),
                    rv.GetProperty("seq").GetInt64(),
                    rv.GetProperty("at").GetInt64());
                Assert.Equal(rv.GetProperty("body_hex").GetString(), Hex(r.Bytes()));
                receipts.Add(r);
                sigs.Add(Cose.MldsaSign(Alg, Seed, r.Bytes())); // a real sig, so the BREAK is what fires
            }
            var ex = Assert.Throws<NaalpException>(() => Audit.VerifyChain(receipts, sigs, Alg, Pk));
            Assert.Equal(cb.GetProperty("expect").GetString(), ex.Kind); // ChainBroken
        }

        [Fact]
        public void VerifyChainDetectsTamperedSignature()
        {
            JsonElement v = Vector();
            JsonElement chain = v.GetProperty("chain");
            JsonElement rv0 = chain.GetProperty("receipts")[0];
            var r = new Audit.Receipt(
                Hb(chain.GetProperty("genesis_prev_hex").GetString()!),
                Hb(rv0.GetProperty("obj_hex").GetString()!),
                0, rv0.GetProperty("at").GetInt64());
            byte[] good = Cose.MldsaSign(Alg, Seed, r.Bytes());
            byte[] bad = (byte[])good.Clone();
            bad[bad.Length - 1] ^= 1;
            var ex = Assert.Throws<NaalpException>(
                () => Audit.VerifyChain(new List<Audit.Receipt> { r }, new List<byte[]> { bad }, Alg, Pk));
            Assert.Equal("ReceiptUnsigned", ex.Kind);
        }

        [Fact]
        public void ConsistentWithAnchor()
        {
            Assert.True(Audit.ConsistentWithAnchor(100, 100));
            Assert.True(Audit.ConsistentWithAnchor(99, 100));
            Assert.False(Audit.ConsistentWithAnchor(101, 100)); // created after ordered
        }

        // ---- equivocation detection (design §8.5) ---------------------------------------------

        private static (JsonElement F, Audit.Receipt A, Audit.Receipt B) ForkReceipts()
        {
            JsonElement v = Vector();
            JsonElement f = v.GetProperty("fork_proof");
            var ra = new Audit.Receipt(
                Hb(f.GetProperty("prev_hex").GetString()!), Hb(f.GetProperty("obj_a_hex").GetString()!),
                f.GetProperty("seq").GetInt64(), f.GetProperty("at").GetInt64());
            var rb = new Audit.Receipt(
                Hb(f.GetProperty("prev_hex").GetString()!), Hb(f.GetProperty("obj_b_hex").GetString()!),
                f.GetProperty("seq").GetInt64(), f.GetProperty("at").GetInt64());
            return (f, ra, rb);
        }

        [Fact]
        public void EquivocationReceiptsMatchOracle()
        {
            (JsonElement f, Audit.Receipt ra, Audit.Receipt rb) = ForkReceipts();
            Assert.Equal(f.GetProperty("body_a_hex").GetString(), Hex(ra.Bytes()));
            Assert.Equal(f.GetProperty("body_b_hex").GetString(), Hex(rb.Bytes()));
            JsonElement eq = Vector().GetProperty("equivocation");
            Assert.Equal(eq.GetProperty("receipt_a").GetProperty("body_hex").GetString(), Hex(ra.Bytes()));
            Assert.Equal(eq.GetProperty("receipt_b").GetProperty("body_hex").GetString(), Hex(rb.Bytes()));
        }

        [Fact]
        public void EquivocationDetected()
        {
            // Two validly-signed receipts by ONE authority at one seq naming DIFFERENT objects: the
            // auditor mints a fork proof (corpus expect == "Equivocation"). A benign first observe
            // returns null. THIS is the mutation-target assertion.
            (JsonElement f, Audit.Receipt ra, Audit.Receipt rb) = ForkReceipts();
            byte[] sigA = Cose.MldsaSign(Alg, Seed, ra.Bytes());
            byte[] sigB = Cose.MldsaSign(Alg, Seed, rb.Bytes());
            var auditor = new Audit.Auditor(Alg, Pk, Hb(f.GetProperty("signer_hex").GetString()!), f.GetProperty("ext_counter").GetInt64());
            Assert.Null(auditor.Observe(ra, sigA)); // first observe is benign
            Audit.ForkProof? fp = auditor.Observe(rb, sigB);
            Assert.NotNull(fp); // a genuine fork at one seq must be detected
            fp!.Verify(Alg, Pk); // the minted proof verifies against the accused key (no throw)
            Assert.Equal("Equivocation", Vector().GetProperty("equivocation").GetProperty("expect").GetString());
        }

        [Fact]
        public void BenignDuplicateIsNotEquivocation()
        {
            (JsonElement f, Audit.Receipt ra, _) = ForkReceipts();
            byte[] sigA = Cose.MldsaSign(Alg, Seed, ra.Bytes());
            var auditor = new Audit.Auditor(Alg, Pk, Hb(f.GetProperty("signer_hex").GetString()!), f.GetProperty("ext_counter").GetInt64());
            Assert.Null(auditor.Observe(ra, sigA));
            Assert.Null(auditor.Observe(ra, sigA)); // an exact duplicate is not a fork
        }

        [Fact]
        public void ObserveRejectsUnsigned()
        {
            (JsonElement f, Audit.Receipt ra, _) = ForkReceipts();
            var auditor = new Audit.Auditor(Alg, Pk, Hb(f.GetProperty("signer_hex").GetString()!), f.GetProperty("ext_counter").GetInt64());
            var ex = Assert.Throws<NaalpException>(() => auditor.Observe(ra, new byte[8]));
            Assert.Equal("ReceiptUnsigned", ex.Kind);
        }

        // ---- fork proof (draft-01 §8.5) -------------------------------------------------------

        [Fact]
        public void ForkProofPreimageMatchesOracle()
        {
            (JsonElement f, Audit.Receipt ra, Audit.Receipt rb) = ForkReceipts();
            Audit.ForkProof fp = Audit.NewForkProof(
                Hb(f.GetProperty("signer_hex").GetString()!), ra, Array.Empty<byte>(), rb, Array.Empty<byte>(),
                f.GetProperty("ext_counter").GetInt64());
            Assert.Equal(f.GetProperty("preimage_hex").GetString(), Hex(fp.Preimage()));
        }

        [Fact]
        public void ForkProofVerifyAcceptsAndFailsClosed()
        {
            (JsonElement f, Audit.Receipt ra, Audit.Receipt rb) = ForkReceipts();
            byte[] sigA = Cose.MldsaSign(Alg, Seed, ra.Bytes());
            byte[] sigB = Cose.MldsaSign(Alg, Seed, rb.Bytes());
            byte[] signer = Hb(f.GetProperty("signer_hex").GetString()!);
            long ext = f.GetProperty("ext_counter").GetInt64();

            Audit.NewForkProof(signer, ra, sigA, rb, sigB, ext).Verify(Alg, Pk); // a valid proof (no throw)

            var same = Assert.Throws<NaalpException>(
                () => Audit.NewForkProof(signer, ra, sigA, ra, sigA, ext).Verify(Alg, Pk));
            Assert.Equal("ForkProofInvalid", same.Kind); // same object named twice

            var unnamed = Assert.Throws<NaalpException>(
                () => Audit.NewForkProof(Array.Empty<byte>(), ra, sigA, rb, sigB, ext).Verify(Alg, Pk));
            Assert.Equal("ForkProofInvalid", unnamed.Kind); // unnamed accused

            var rbSeq = new Audit.Receipt(rb.Prev, rb.Obj, rb.Seq + 1, rb.At);
            byte[] sigB2 = Cose.MldsaSign(Alg, Seed, rbSeq.Bytes());
            var mism = Assert.Throws<NaalpException>(
                () => Audit.NewForkProof(signer, ra, sigA, rbSeq, sigB2, ext).Verify(Alg, Pk));
            Assert.Equal("ForkProofInvalid", mism.Kind); // seq mismatch

            byte[] badB = (byte[])sigB.Clone();
            badB[badB.Length - 1] ^= 1;
            var tampered = Assert.Throws<NaalpException>(
                () => Audit.NewForkProof(signer, ra, sigA, rb, badB, ext).Verify(Alg, Pk));
            Assert.Equal("ReceiptUnsigned", tampered.Kind); // a signature that does not verify
        }

        // ---- offline causal graph (design §8.2-§8.3) ------------------------------------------

        private static List<Audit.CausalNode> Nodes(string key)
        {
            var nodes = new List<Audit.CausalNode>();
            foreach (JsonElement n in Vector().GetProperty(key).GetProperty("nodes").EnumerateArray())
            {
                var causes = new List<byte[]>();
                foreach (JsonElement c in n.GetProperty("causes_hex").EnumerateArray())
                {
                    causes.Add(Hb(c.GetString()!));
                }
                nodes.Add(new Audit.CausalNode(Hb(n.GetProperty("id_hex").GetString()!), causes, n.GetProperty("position").GetInt64()));
            }
            return nodes;
        }

        [Fact]
        public void CausalValidTopoOrderMatchesOracle()
        {
            List<Audit.CausalNode> nodes = Nodes("causal_valid");
            Audit.VerifyCausal(nodes); // no throw
            List<byte[]> order = Audit.TopoOrder(nodes);
            var want = new List<string>();
            foreach (JsonElement e in Vector().GetProperty("causal_valid").GetProperty("topo_order_hex").EnumerateArray())
            {
                want.Add(e.GetString()!);
            }
            var got = new List<string>();
            foreach (byte[] o in order)
            {
                got.Add(Hex(o));
            }
            Assert.Equal(want, got);
        }

        [Fact]
        public void CausalCycleRejected()
        {
            JsonElement v = Vector();
            var ex = Assert.Throws<NaalpException>(() => Audit.VerifyCausal(Nodes("causal_cycle")));
            Assert.Equal(v.GetProperty("causal_cycle").GetProperty("expect").GetString(), ex.Kind);
        }

        [Fact]
        public void CausalFutureCauseRejected()
        {
            JsonElement v = Vector();
            var ex = Assert.Throws<NaalpException>(() => Audit.VerifyCausal(Nodes("causal_future")));
            Assert.Equal(v.GetProperty("causal_future").GetProperty("expect").GetString(), ex.Kind);
        }
    }
}
